module/Finna/src/Finna/RecordDriver/SolrEad3.php

```diff
@@ -450,81 +451,157 @@ public function archiveRequestAllowed()
+     * @param bool  $translations Include only authors with roles that have translations
+     * @param bool  $unknown      Include only authors with unknown roles or roles without translations
+     *
+     * @return array
+     */
+    public function getAuthors(
+        array $exclude = [],
+        bool $translations = false,
+        bool $unknown = false
+    ): array {
+        $result = [];
+        // Check authors under controlaccess
+        $names = $this->getAuthorElements();
         foreach ($names as $node) {
-            $name = $this->getDisplayLabel($node, 'part', true);
-            if (empty($name) || !$name[0]) {
+            $attr = $node->attributes();
+            $localtype = trim((string)($attr->localtype ?? ''));
+            $relator = (string)$attr->relator;
+            $relatorLC = mb_strtolower($relator, 'UTF-8');
+            $role = $this->translateRole($localtype) ?? $this->translateRole($relatorLC);
+            if (!empty($exclude) && in_array($relatorLC, $exclude)) {
+                continue;
+            }
+            if ((!$role && $translations) || ($role && $unknown)) {
                 continue;
             }
-            $result[] = ['name' => $name[0]];
-        }
-        if (!empty($result)) {
-            return $result;
-        }
-
-        // Not found, search again without language filters
-        foreach ($names as $node) {
             $name = $this->getDisplayLabel($node);
             if (empty($name) || !$name[0]) {
                 continue;
             }
-            $result[] = $name[0];
+            $result[] = [
+                'id' => (string)$node->attributes()->identifier,
+                'role' => $role ?: $relatorLC,
+                'name' => $name[0],
+            ];
+        }
+        // Check authors under relations
+        $relations = $this->getRelations('cpfrelation', $exclude);
+        foreach ($relations as $relation) {
+            $role = $this->translateRole($relation['role']);
+            if ((!$role && $translations) || ($role && $unknown)) {
+                continue;
+            }
+            if ($role) {
+                $relation['role'] = $role;
+            }
+            // Do not add duplicates
+            foreach ($result as $res) {
+                if ($res === $relation) {
+                    continue 2;
+                }
+            }
+            $result[] = $relation;
         }
-        $result = array_map(
-            function ($name) {
-                return ['name' => $name];
-            },
-            array_unique($result)
-        );
-
         return $result;
     }
 
     /**
      * Get relations.
      *
+     * @param string $relationtype Relationtype to be included
+     * @param array  $exclude      Arcroles to be excluded
+     *
      * @return array
      */
-    public function getRelations()
-    {
+    public function getRelations(
+        string $relationtype,
```

Comment:
Tämä ei voi alkaa vaatia parametria yhtäkkiä, ei ole enää yhteensopiva muiden formaattien kanssa. Ehkä olisi parempi olla protected:na getRelationsWithType tms. jota getRelations ja muut käyttäisivät.

Reply:
SolrAuthEaccpf:n ja SolrAuthMarcin getRelations on tarkoitukseltaan vähän eri, ehkä SolrEad3ssa ei ees tarvii vastaavaa, jos vaan jätän pois? Tai ei sitä ainakaan nyt käytettäis missään kun muutin tuon nimen

Reply:
Ehkä joo. Toivotaan niin! :)